no keyring and identity panic

# 1. Build defradb
  `go build -o /tmp/defradb-bin ./cmd/defradb`

  # 2. Start with local ACP and no keyring
 ` /tmp/defradb-bin start --no-keyring --rootdir /tmp/defradb-repro --document-acp-type local`

  # 3. In another terminal, add a policy
 ` /tmp/defradb-bin client --no-keyring --rootdir /tmp/defradb-repro \
    acp document policy add -f examples/policy/dac_policy.yml \
    -i e3b722906ee4e56368f581cd8b18ab0f48af1ea53e635e3f7b8acd076676f6ac`

  # 4. Add a schema linked to the policy (use the PolicyID from step 3)
  `/tmp/defradb-bin client --no-keyring --rootdir /tmp/defradb-repro \
    schema add 'type Users @policy(id: "<PolicyID from step 3>", resource: "users") { name: String }'`

  # 5. Add a document WITH identity - this triggers the panic
  `/tmp/defradb-bin client --no-keyring --rootdir /tmp/defradb-repro \
    collection add --name Users '{"name": "Secret"}' \
    -i e3b722906ee4e56368f581cd8b18ab0f48af1ea53e635e3f7b8acd076676f6ac`

  Step 5 returns Error: 500: and the server logs show:

 ```
panic: runtime error: invalid memory address or nil pointer dereference

  -> github.com/sourcenetwork/defradb/internal/core/block.signBlock
  ->   internal/core/block/signing.go:73
```

  The panic is at ident.PrivateKey().Type() - when --no-keyring is used, the identity has no private key, so
  PrivateKey() returns nil. Adding a document without --identity (step 4, schema add) works fine because no
  signing is attempted. The crash only happens when both --no-keyring and --identity are used together on a
  write operation.

--- a/internal/core/block/signing.go
+++ b/internal/core/block/signing.go
@@ -44,7 +44,7 @@ func extractFullIdentity(ctx context.Context) immutable.Option[identity.FullIden
 	}
 
 	fullIdent, ok := ident.Value().(identity.FullIdentity)
-	if !ok {
+	if !ok || fullIdent.PrivateKey() == nil {
 		return immutable.None[identity.FullIdentity]()
 	}
 

--- a/internal/core/block/signing_test.go
+++ b/internal/core/block/signing_test.go
@@ -1,0 +1,80 @@
+// Copyright 2026 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package coreblock
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/sourcenetwork/defradb/acp/identity"
+	"github.com/sourcenetwork/defradb/crypto"
+	iIdentity "github.com/sourcenetwork/defradb/internal/identity"
+	"github.com/sourcenetwork/immutable"
+)
+
+func TestEnabledSigningFromContext_WithFullIdentity_ReturnsIdentity(t *testing.T) {
+	ident, err := identity.Generate(crypto.KeyTypeSecp256k1)
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	ctx = iIdentity.WithContext(ctx, immutable.Some[identity.Identity](ident))
+	ctx = ContextWithEnabledSigning(ctx)
+
+	enabled, fullIdent := EnabledSigningFromContext(ctx)
+	require.True(t, enabled)
+	require.True(t, fullIdent.HasValue())
+	require.NotNil(t, fullIdent.Value().PrivateKey())
+}
+
+func TestEnabledSigningFromContext_WithTokenIdentityNilPrivateKey_ReturnsNone(t *testing.T) {
+	// Generate an identity and create a bearer token from it.
+	ident, err := identity.Generate(crypto.KeyTypeSecp256k1)
+	require.NoError(t, err)
+	err = ident.UpdateToken(time.Hour, immutable.Some("test"), immutable.None[string]())
+	require.NoError(t, err)
+
+	// Parse the token back, this creates a FullIdentity with nil PrivateKey.
+	// This is exactly what happens on the server side when a client sends a
+	// request with a bearer token (e.g. --no-keyring with --identity).
+	tokenIdent, err := identity.FromToken([]byte(ident.BearerToken()))
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	ctx = iIdentity.WithContext(ctx, immutable.Some[identity.Identity](tokenIdent))
+	ctx = ContextWithEnabledSigning(ctx)
+
+	enabled, fullIdent := EnabledSigningFromContext(ctx)
+	require.True(t, enabled)
+
+	// If extractFullIdentity returns Some(fullIdent) with nil PrivateKey, that can
+	// lead to a panic in signBlock. After the fix, it returns None so signing is skipped.
+	require.False(t, fullIdent.HasValue())
+}
+
+func TestEnabledSigningFromContext_WithNoIdentity_ReturnsNone(t *testing.T) {
+	ctx := context.Background()
+	ctx = ContextWithEnabledSigning(ctx)
+
+	enabled, fullIdent := EnabledSigningFromContext(ctx)
+	require.True(t, enabled)
+	require.False(t, fullIdent.HasValue())
+}
+
+func TestEnabledSigningFromContext_WithSigningDisabled_ReturnsFalse(t *testing.T) {
+	ctx := context.Background()
+
+	enabled, fullIdent := EnabledSigningFromContext(ctx)
+	require.False(t, enabled)
+	require.False(t, fullIdent.HasValue())
+}

